url969d1ddca4_compmstr_java_scheme_tgz-pJ8-com_undi_javascheme_JavaSchemeJ8/src/com/undi/javascheme/SchemeReader.java

```diff
@@ -0,0 +1,427 @@
+      }
+      break;
+    }
+    peekExpectedDelimiter();
+    return SchemeObject.makeCharacter((short) c);
+  }
+  
+  public SchemeObject readBoolean(){
+    //char[] boolString = readToken().toCharArray();
+    eatWhitespace();
+    char boolChar = (char)getc();
+    boolean retVal;
+    if(boolChar == 't'){
+      retVal = true;
+    }else if(boolChar == 'f'){
+      retVal = false;
+    }else{
+      retVal = false;
+			throw new SchemeException("Invalid boolean: " + boolChar);
+    }
+    return SchemeObject.makeBoolean(retVal);
+  }
+  
+  public SchemeObject readPair(){
+    eatWhitespace();
+    int c = getc();
+    if(c == ')'){
+      return SchemeObject.THE_EMPTY_LIST;
+    }
+    ungetc(c);
+    SchemeObject carObject = read();
+    
+    SchemeObject cdrObject = null;
+    eatWhitespace();
+    c = getc();
+    if(c == '.'){
+      //take in the improper list form
+      c = peek();
+      if(!isDelimiter(peek())){
+				throw new SchemeException("Error: dot not followed by delimiter");
+      }
+      cdrObject = read();
+      eatWhitespace();
+      c = getc();
+      if(c != ')'){
+        throw new SchemeException("Error: no trailing right paren on dot form");
+      }
+    }else{
+      ungetc(c);
+      cdrObject = readPair();
+    }
+    
+    return SchemeObject.cons(carObject, cdrObject);
+  }
+  
+  public SchemeObject readSymbol(){
+    StringBuilder buffer = new StringBuilder();
+    
+    int c = getc();
+    while(isInitial(c) || Character.isDigit(c) ||
+        c == '+' || c == '-' || c == '.' || c == '/' || c == '_' || c == '&'){
+      buffer.append((char)c);
+      c = getc();
+    }
+    if(isDelimiter(c)){
+      ungetc(c);
+      return SchemeObject.makeSymbol(buffer.toString());
+    }else{
+      throw new SchemeException("Symbol not followed by delimiter");
+    }
+  }
+  
+  public SchemeObject readKeyword(){
+	  SchemeObject sym = readSymbol();
+	  return SchemeObject.makeKeyword(sym.getSymbol());
+  }
+  
+  /**
+   * 
+   * @return the object read, or null if end of file is reached
+   */
+  public SchemeObject read(){
+    boolean quoted = false;
+    eatWhitespace();
+    if(peek() == '\''){
+      getc();
+      quoted = true;
+    }
+    SchemeObject.type nextType = nextType();
+    if(nextType == null){
+      //End of file reached, return null
+      return null;
+    }
+    if(this.readFuncs[nextType.ordinal()] != null){
+      if(quoted){
+        return SchemeObject.cons(SchemeObject.QUOTE_SYMBOL, SchemeObject.cons(this.readFuncs[nextType.ordinal()].read(), SchemeObject.THE_EMPTY_LIST));
+      }else{
+        return this.readFuncs[nextType.ordinal()].read();
+      }
+    }
+
+    throw new SchemeException("Unsupported input type in read");
+  }
+  
+  /**
+   * Determines what the next SchemeObject type will be from the start of the stream
+   * @return the next type of objec to read, or null if end of file
+   */
+  private SchemeObject.type nextType(){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.